File: ADT/DSU/detect_cycle_using__dsu.py
from typing import List

class Solution:
    def __init__(self):
        self.parent = []
        self.rank = []

    def find(self, x: int) -> int:
    	if self.parent[x] == x:
    		return x
    	# path compression by caching
    	self.parent[x] = self.find(self.parent[x])
    	return self.parent[x]
        
    def Union(self, x: int, y: int) -> None:
    	x_parent = self.find(x)
    	y_parent = self.find(y)

    	# union find by rank
    	if self.rank[x_parent] > self.rank[y_parent]:
    		self.parent[y_parent] = x_parent
    	elif self.rank[x_parent] < self.rank[y_parent]:
    		self.parent[x_parent] = y_parent
    	else:
    		self.parent[x_parent] = y_parent
    		self.rank[y_parent] += 1

    # Function to detect cycle using DSU in an undirected graph.
    def detectCycle(self, V: int, adj: List[List[int]]) -> bool:
    	self.parent = list(range(V))
    	self.rank = [0] * V

    	for u in range(V):
    		for v in adj[u]:
    			if u < v:
    				if self.find(u) == self.find(v):
    					return True

    				else:
    					self.Union(u, v)
    	return False

File: ADT/DSU/detect_cycle_using__dsu.js
class Solution {
    constructor() {
        this.parent = [];
        this.rank = [];
    }

    find(x) {
        if (this.parent[x] === x)
            return x;

        this.parent[x] = this.find(this.parent[x]);
        return this.parent[x];
    }

    Union(x, y) {
        const x_parent = this.find(x);
        const y_parent = this.find(y);

        if (this.rank[x_parent] > this.rank[y_parent]) {
            this.parent[y_parent] = x_parent;
        } else if (this.rank[x_parent] < this.rank[y_parent]) {
            this.parent[x_parent] = y_parent;
        } else {
            this.parent[x_parent] = y_parent;
            this.rank[y_parent]++;
        }
    }

    // Function to detect cycle using DSU in an undirected graph.
    detectCycle(V, adj) {
        this.parent = new Array(V);
        this.rank = new Array(V).fill(0);

        for (let i = 0; i < V; i++)
            this.parent[i] = i;

        for (let u = 0; u < V; u++) {
            for (const v of adj[u]) {
                if (u < v) {
                    if (this.find(u) === this.find(v))
                        return true;
                    else {
                        this.Union(u, v);
                    }
                }
            }
        }
        return false;
    }
}